pkg/knode-manager/controllers/secret.go

```diff
@@ -0,0 +1,169 @@
+	klog.Info("enqueue secret delete ", "key ", key)
+}
+
+func (s *SecretController) UpdateSecret(oldObj, newObj interface{}) {
+	key, err := cache.MetaNamespaceKeyFunc(newObj)
+	if err != nil {
+		runtime.HandleError(err)
+		return
+	}
+	s.secretQueue.Add(key)
+	klog.Info("secret update ", "key ", key)
+}
+
+// Run starts and listens on channel events
+func (s *SecretController) Run(ctx context.Context) error {
+	stopCh := ctx.Done()
+
+	defer s.secretQueue.ShutDown()
+	defer s.secretQueue.ShutDown()
+	klog.Infof("kosmos starting secret controller")
+	defer klog.Infof("kosmos shutting secret controller")
+	if !cache.WaitForCacheSync(stopCh, s.synced) {
+		return fmt.Errorf("kosmos cannot secret secret cache")
+	}
+	klog.Infof("kosmos sync secret cache success")
+
+	go wait.Until(s.syncSecret, 0, stopCh)
+
+	<-stopCh
+	return nil
+}
+
+func (s *SecretController) syncSecret() {
+	secret2key, result := s.secretQueue.Get()
+	if result {
+		return
+	}
+	defer s.secretQueue.Done(secret2key)
+	ns, secretName, err := cache.SplitMetaNamespaceKey(secret2key.(string))
+	if err != nil {
+		s.secretQueue.Forget(secret2key.(string))
+		return
+	}
+
+	klog.Infof("kosmos sync secret: start sync secret, secret: %v", secret2key.(string))
+	secretLifecycle := false
+
+	secretInstance, err := s.masterLister.Secrets(ns).Get(secretName)
+	if err != nil {
+		if !apierrors.IsNotFound(err) {
+			klog.Errorf("kosmos sync: get secret from list failed, error: %v", err)
+			return
+		}
+		secretLifecycle = true
+	}
+	defer func() {
+		if err != nil {
+			s.secretQueue.AddRateLimited(secret2key)
+			return
+		}
+		s.secretQueue.Forget(secret2key)
+	}()
+	if secretLifecycle || secretInstance.DeletionTimestamp != nil {
+		err = s.handler.Delete(ns, secretName)
+		if err != nil {
+			klog.Errorf("kosmos sync: delete secret from downstream list failed, secret: %v", secretName)
+			return
+		}
+		klog.Infof("kosmos sync: delete secret from downstream list succeed, secret: %v", secretName)
+		return
+	}
+
+	klog.Infof("Secret %v/%v to be update", ns, secretName)
+	s.syncSecretHandler(secretInstance)
+}
+
+func (s *SecretController) syncSecretHandler(secret *corev1.Secret) {
+	key, err := cache.MetaNamespaceKeyFunc(secret)
+	if err != nil {
+		runtime.HandleError(err)
+		return
+	}
+	defer func() {
+		if err != nil {
+			klog.Error(err)
+			s.secretQueue.AddRateLimited(key)
+			return
+		}
+	}()
+	var secretInSub *corev1.Secret
+	secretInSub, err = s.clientLister.Secrets(secret.Namespace).Get(secret.Name)
+	if err != nil {
+		klog.Infof("secretInSub %v err: %v", secretInSub, err)
+		return
+	}
+
+	secretCopy := secret.DeepCopy()
+	if err = utils.FilterSecret(secretCopy); err != nil {
+		return
+	}
+	klog.Infof("Old secret %+v\n, new secret %+v", secretInSub, secretCopy)
+	if _, err = s.handler.Update(secretInSub, secretCopy); err != nil {
+		klog.Infof("Update secret %v err: %v", secretInSub, err)
+		return
+	}
+	klog.Infof("Handler secret: finished processing %q", secret.Name)
```

Comment:
i didn't see Create event to be handled

Reply:
The create event already exists in the pod controller.

Reply:
ok, i see see